### ISSUE TITLE
GoogleOptimizeRouteService - Enhance - Retry Logic

### ISSUE THREAD

**Issue by stevejgordon** [HTBox/allReady#1576]
Consider adding retry logic within the optimize route service so that a failed call to Google is retried up to 3 times to account for any service outages.

While this may add resiliency it would add time to the process and this may not be what we want to do.

@tonysurma @MisterJames - Open to your thoughts on this?

**Comment by mgmccarthy** [comment]
@stevejgordon is there any way to communicate the data back asynchronously to the user who issued this request that invokes the Google service that is not something directly tied to displaying something in the UI?

If we absolutely need to present this on the UI, then I'd say keep the retries on the low side. 

If there is a way, however, to communicate to the user who is issuing the request that will invoke Google service the result of the Google call another way (examples are email/sms), then you can probably get a better retry policy and avoid RPC locks.

**Comment by EmilyLuijbregts** [comment]
@stevejgordon - Just doing a follow up on this Steve. Have you had chance to look at the above question?
@tonysurma and @MisterJames - just in case you guys wanted to add your thoughts in here :)

**Comment by stevejgordon** [comment]
@mgmccarthy - In this case the user is initiating a request to re-order the requests and I therefore think waiting on the response and updating the UI makes sense. So I'm in agreement, retying too many times might not be great. I think it's reasonable to expect some delay since this is an explicit operation but then again it's not critical.

I think I'd rather investigate an idea I have around a simple messaging flow so that when the UI updates I can provide further details about the success or failure of the optimize route attempt. That way, if it fails, it fails fast but the user is aware and can retry themselves.

We should also get some logging put in place when we can so that when we go live, failed calls can be recorded and reviewed. It's reasonable to expect the odd issue calling the Google API, but if we see repeated problems we have something to investigate.

**Comment by mgmccarthy** [comment]
@stevejgordon, looking at my previous reply, I realize that I didn't ask what I should have which is when the GoogleOptimizeRouteService is invoked, is this during a POST or a GET?

Also, is there a controller/view name that was added where I could try to run this locally so I could learn how this works? Thanks!

**Comment by stevejgordon** [comment]
@mgmccarthy - So, you'll need to have an itinerary which once created you'd need to edit, to setup a start/end address for it. You'd also need some requests assigned to it with appropriate lat/long co-ordinates. With that all in place, if you are on the itinerary details screen you should see some links appear to "optimize route" etc. Clicking optimize route will show a confirm modal and once you confirm that, it will post to our action, which will call a handler which calls off to Google API.

The starting point if you want to read through the code flow is on the ItineraryController OptimizeRoute action.

**Comment by mgmccarthy** [comment]
@stevejgordon, my two cents on this issue after looking at `ItineraryController.OptimizeRoute` action method and `OptimizeRouteCommandHandler`. IMHO...

In `OptimizeRoutecommandHandler` we're invoking both a a non-transactional resource (the Google API) and transactional resource (the db) and in the same logical transaction. Although MediatR logically separates handlers for us, physically, it treats the controller/command(s)/notification(s) as one transactional unit.

So even though we don't get that nice physical transactional separation from MediatR (like we would from a service bus), we can still logically separate the two concerns in our code.

After the read from the Google API, you could publish a notification, `RouteOptimized`, then do the work of re-ordering the `ItineraryRequests` in a separate handler. This gives us a nice separation of concerns, plus, it could set up other areas of the system that might want to know when we optimize a route by simply subscribing to `RouteOptimized`.

Also, I don't know if there is any validation you can front in `ItineraryController.OptimizeRoute` before sending `OptimizeRouteCommand`, but any validation we can do before sending a command that the command handler does not have to do is a good thing.

We want to send commands that have a very high chance of succeeding, and from a pure CQRS approach, command handlers should never return anything. In AllReady's case, we need to return entity id's b/c the database owns pk generation not our code, but besides that, we don't really want our command handlers returning anything.

If you check out `RequestApiController` in PR #1598, you'll see I validate up front, then send the command with no return value. It is a little more work to do this sometimes, as we need to write more Query/QueryHandlers, but it keeps our commands focused on "doing" something, not validating a bunch of stuff, then doing it.

Again, this is IMHO, but I figured I'd chime in.

**Comment by stevejgordon** [comment]
@mgmccarthy Great points and I like where you're going with it. I do have a clarification though. My understanding at a high level was that notifications are handled without the callers actually awaiting the work. I've not actually tested that but that was my assumption as we could have multiple handlers for the notification which would run in the background. I think I'm reading that you don't believe that is the case?

My main concern is that the controller will need to know that the database has been updated, before the redirect so that when the details page loads it shows any newly ordered requests. In my naive assumption of how notifications work, that wouldn't be the case as I'd assume the command would come back as seen as the notification was sent and we might reload the page prior to the db work completing.

**Comment by mgmccarthy** [comment]
@stevejgordon, I want to update my initial feedback above.

**First**, since we're doing a "read" from the `_optimizeRouteService` (essentially, a GET), and then updating the db in the same handler, then the handler is good as-is. Writing this comment, I was fresh off of building HttpClient invocation for GASA POST back to their API. Since the concept of `RouteOptimized` basically means updating the ItineraryRequest's order, then this handler is good to go. 

That being said, it looks like there are a couple different places in the system where we're re-ordering `ItineraryRequet`'s (`ReorderRequestCommandHandler`, `RemoveRequestCommandHandler`, `AddRequestsToItineraryCommandHandler`). 

If find ourselves changing just the `OrderIndex` property separate from the rest of the data on `ItineraryRequest` much more often than all the data on ItineraryRequest together, then to me, from a transactional point of view, the OrderIndex data should be broken out into a separate model, `ItineraryRequestOrderIndex`, and we can scope the transaction for updating OrderIndex to a smaller set of data. 

Granted, this introduces more work on reads (extra Include), and for UI's where we write all the data for an "Itinerary Request" (`ItineraryRequest`/`ItineraryRequestOrderIndex`) at the same time we have to dispatch two commands from the controller, b/c each handler would be transactionally scoped to its model, but this aligns our data storage with what we need to be transactionally consistent,

One more thing about this point, is if we have multiple pieces of code doing the same thing (re-ordering ItineraryRequest), then maybe it's possible to bring this code to a single handler, and have that handler do the work of re-ordering the ItineraryRequests? Just a thought.

Also, I'm thinking any of this work is post v1 at this point, as it would represent a good amount of refactoring.

**Second**, I did not know that notifications worked this way in MediatR? It was my understanding that MediatR is sync all the way. 

I can put a quick test together of a controller/command handler/notification handlers and see if control returns to the UI any time after the notification is dispatched from the command handler/is handled by the notification handlers. If it is not (which is how I understand MediatR to work), then this concern:
> My main concern is that the controller will need to know that the database has been updated, before the redirect so that when the details page loads it shows any newly ordered requests

should not longer be a concern.

Looking at the GitHub page, it does not really get into async dispatch of notifications, especially if there is more than one notification handler involved.

**Comment by stevejgordon** [comment]
@mgmccarthy - Cool. Sounds like we're thinking along the same lines here. I certainly feel having a single responsible service/handler for reordering could well be a good refactor for the future.

I think in regards to this specific issue, the retry is probably not needed. I agree that trying to prevalidate the data should happen, any 0,0 lat/longs for example are unlikely to be valid and would just throw off the result.

Adding some basic failure messaging would be best I think so that if we can't re-order the user knows and can try again at will.

Regarding the notifications, I'm purely assuming at this stage. I have a small project at home where I've used them and I could probably test easily by adding in a fake delay on the handler and watching to see if that's reflected in the UI. Would be good to know for sure.

**Comment by mgmccarthy** [comment]
> I think in regards to this specific issue, the retry is probably not needed.

If we eventually need the ItineraryRequests re-ordered, I think you could do the retry and keep it on the low side. Sure, the UI will "wait", but if the error is transient, a Polly retry policy **might** be able to overcome that.

The real question here is how do we "let the user know to try again/how do they try again"? We would return to them on the UI that the optimize route failed and just let them click the button again? It's the compensating action we're requiring the user to take that would need to be defined here. How many times would the user by patient enough to retry this operation? Also, what are the ramifications to the system if the service was down and we could not optimize route for an hour, a day, etc...

> Regarding the notifications, I'm purely assuming at this stage.

https://github.com/jbogard/MediatR/issues/112

we'll see if Jimmy (or someone else in the repo) gets back to me.

**Comment by mgmccarthy** [comment]
I've opened a new issue about making any web service calls (in this case, to Google, async using azure if we can). This is a non-v1 item, but figured I'd just dump the Issuein here for reference:

Issues #1639 and #1626

**Comment by stevejgordon** [comment]
Closing as I've added Polly to cover this off already.
